### PR TITLE
[Java feature server] Converge ServingService API to make Python and Java feature servers consistent

### PR DESCRIPTION
**Pull request by pyalex**
**What this PR does / why we need it**:

Currently, Java Feature Server and Python Feature Server provide two different client APIs. This PR aimed to fix that.

1. Changes to proto API: use `GetOnlineFeatureRequest` (that was created for and currently used by Python FS) as input in `GetOnlineFeatures` method. That also means that Java FS will support requests by **feature service**.
2. `GetOnlineFeatureResponseV2` is proposed as optimized version of `GetOnlineFeatureResponse` and was designed symmetrically to `GetOnlineFeatureRequest`. According to my evaluation serialization/deserialization of the new message is 3x times faster (on 100 rows). That difference increases with a growing number of rows and features. New message also optimized for compaction (message size), which will lead to faster network transmission.

What was updated:
1. Java Feature Server
2. Java SDK
3. Go SDK

What's not covered in this PR:
1. Python FS to return new response

Some clean up:
* "Feature Table" is replaced to "Feature View"
* Suffix "V<N>" is dropped when there are no other versions

**Which issue(s) this PR fixes**:

Fixes #

**Does this PR introduce a user-facing change?**:

```release-note
Java feature server supports retrieving features by feature service name.

ServingService proto API is refactored:
GetOnlineFeaturesResponse now has new argument `request_context` for passing input to ODFVs (previously this inputs were mixed into `entities`
GetOnlineFeaturesResponseV2 is refactored version of response optimized for fast serialization / deserialization and high compaction
```

### CODE REVIEW

**Comment by pyalex** [comment]
Benchmarks were also added into Java integration tests
```
Test 100 rows; Full FS took (min): 32 ms
Test 100 rows; Full FS took (avg): 40.117172 ms
Test 100 rows; Full FS took (median): 36.000000 ms
Test 100 rows; Full FS took (95p): 64.000000 ms
Test 100 rows; Full FS took (99p): 82.000000 ms

Test 100 rows; 50 features took (min): 17 ms
Test 100 rows; 50 features took (avg): 20.864646 ms
Test 100 rows; 50 features took (median): 19.000000 ms
Test 100 rows; 50 features took (95p): 28.000000 ms
Test 100 rows; 50 features took (99p): 55.000000 ms

Test 100 rows; 10 features took (min): 3 ms
Test 100 rows; 10 features took (avg): 5.576768 ms
Test 100 rows; 10 features took (median): 5.000000 ms
Test 100 rows; 10 features took (95p): 15.000000 ms
Test 100 rows; 10 features took (99p): 18.000000 ms

Test 100 rows; 100 features took (min): 17 ms
Test 100 rows; 100 features took (avg): 20.643434 ms
Test 100 rows; 100 features took (median): 19.000000 ms
Test 100 rows; 100 features took (95p): 26.000000 ms
Test 100 rows; 100 features took (99p): 55.000000 ms
```

Full FS == feature service with 250 features

File: protos/feast/serving/ServingService.proto [hunk 1]
```diff
@@ -107,35 +102,40 @@ message GetOnlineFeaturesResponse {
...
-        // This could suggest that the feature values have not yet been ingested
-        // into feast or the ingestion failed.
-        NOT_FOUND = 3;
-
-        // Values could be found for entity key, but field values are outside the maximum 
-        // allowable range.
-        OUTSIDE_MAX_AGE = 4;
+}
+
+message GetOnlineFeaturesResponseV2 {
```
Comment (adchia) [inline protos/feast/serving/ServingService.proto:107]: are you keeping the old response for backwards compatibility or something? What's the reason to keep `GetOnlineFeaturesResponse`?

File: java/sdk/java/src/main/java/dev/feast/FeastClient.java [hunk 2]
```diff
@@ -118,11 +117,56 @@ public GetFeastServingInfoResponse getFeastServingInfo() {
...
    */
-  public List<Row> getOnlineFeatures(List<String> featureRefs, List<Row> rows) {
-    return getOnlineFeatures(featureRefs, rows, "");
+  public List<Row> getOnlineFeatures(List<String> featureRefs, List<Row> entities) {
+    GetOnlineFeaturesRequest.Builder requestBuilder = GetOnlineFeaturesRequest.newBuilder();
+
+    requestBuilder.setFeatures(
+        ServingAPIProto.FeatureList.newBuilder().addAllVal(featureRefs).build());
+
+    Map<String, ValueProto.RepeatedValue.Builder> columnarEntities = new HashMap<>();
```
Comment (adchia) [inline java/sdk/java/src/main/java/dev/feast/FeastClient.java:129]: maybe abstract 129-137 out into a getEntityValuesMap() method? and then you can put all the entities at once

File: protos/feast/serving/ServingService.proto [hunk 3]
```diff
@@ -107,35 +102,40 @@ message GetOnlineFeaturesResponse {
...
-enum FeastServingType {
-    FEAST_SERVING_TYPE_INVALID = 0;
-    // Online serving receives entity data directly and synchronously and will
-    // respond immediately.
-    FEAST_SERVING_TYPE_ONLINE = 1;
-    // Batch serving receives entity data asynchronously and orchestrates the
-    // retrieval through a staging location.
-    FEAST_SERVING_TYPE_BATCH = 2;
+message GetOnlineFeaturesResponseMetadata {
+    FeatureList feature_names = 1;
```
Comment (adchia) [inline protos/feast/serving/ServingService.proto:119]: any reason why we have a FeatureList wrapper?

File: protos/feast/serving/ServingService.proto [hunk 4]
```diff
@@ -107,35 +102,40 @@ message GetOnlineFeaturesResponse {
...
-        NOT_FOUND = 3;
-
-        // Values could be found for entity key, but field values are outside the maximum 
-        // allowable range.
-        OUTSIDE_MAX_AGE = 4;
+}
+
+message GetOnlineFeaturesResponseV2 {
+    GetOnlineFeaturesResponseMetadata metadata = 1;
+    repeated FeatureVector results = 2;
```
Comment (adchia) [inline protos/feast/serving/ServingService.proto:109]: maybe add a comment here that the expectation is that the featurelist length in the metadata should match the number of feature vectors?

File: java/serving/src/test/resources/docker-compose/feast10/materialize.py [hunk 5]
```diff
@@ -56,8 +56,55 @@
...
+    features = [f"feature_{i}" for i in range(num_features)]
+    columns = ["entity", "event_timestamp"] + features
+    df = pd.DataFrame(0, index=np.arange(num_rows), columns=columns)
+    df["event_timestamp"] = datetime.utcnow()
+    for column in ["entity"] + features:
+        df[column] = np.random.randint(1, key_space, num_rows)
+
+    df.to_parquet(destination)
+
+generate_data(10**3, 250, 10**3, "benchmark_data.parquet")
```
Comment (adchia) [inline java/serving/src/test/resources/docker-compose/feast10/materialize.py:72]: maybe link to the RFC?

File: java/serving/src/test/java/feast/serving/it/ServingBenchmarkIT.java [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+    Stopwatch timer = Stopwatch.createUnstarted();
+
+    List<Long> records = Lists.newArrayList();
+
+    for (int i = 0; i < runs; i++) {
+      timer.reset();
+      timer.start();
+      target.run();
+      timer.stop();
+      if (i >= WARM_UP_COUNT) {
```
Comment (adchia) [inline java/serving/src/test/java/feast/serving/it/ServingBenchmarkIT.java:132]: seems fishy that you're doing your own warmups. Maybe consider JMH to manage the benchmarks?

File: protos/feast/serving/ServingService.proto (hunk 1 above)
Comment (pyalex) [reply protos/feast/serving/ServingService.proto:107]: It's still being used by Python implementation of feature server and should be removed when python part will be migrated

File: protos/feast/serving/ServingService.proto (hunk 3 above)
Comment (pyalex) [reply protos/feast/serving/ServingService.proto:119]: I guess this is just symmetrical to GetOnlineFeatureRequest

File: java/serving/src/test/java/feast/serving/it/ServingBenchmarkIT.java (hunk 6 above)
Comment (pyalex) [reply java/serving/src/test/java/feast/serving/it/ServingBenchmarkIT.java:132]: I spent an hour trying to make JMH annotations work together with junit in maven and I failed. I think it's not worth it. I wrote this code in a less time

File: protos/feast/serving/ServingService.proto [hunk 7]
```diff
@@ -94,6 +87,8 @@ message GetOnlineFeaturesRequest {
     // A map of entity name -> list of values
     map<string, feast.types.RepeatedValue> entities = 3;
     bool full_feature_names = 4;
+
+    map<string, feast.types.RepeatedValue> request_context = 5;
 }
```
Comment (adchia) [inline protos/feast/serving/ServingService.proto:92]: Maybe add comments on this too?

File: sdk/go/request.go [hunk 8]
```diff
@@ -29,24 +30,38 @@ type OnlineFeaturesRequest struct {
...
 	if err != nil {
 		return nil, err
 	}
 
 	// build request entity rows from native entities
-	entityRows := make([]*serving.GetOnlineFeaturesRequestV2_EntityRow, len(r.Entities))
-	for i, entity := range r.Entities {
-		entityRows[i] = &serving.GetOnlineFeaturesRequestV2_EntityRow{
-			Fields: entity,
+	entityColumns := make(map[string][]*types.Value, len(r.Entities[0]))
```
Comment (adchia) [inline sdk/go/request.go:40]: why is there an index[0]?

File: java/serving/src/main/java/feast/serving/registry/RegistryRepository.java [hunk 9]
```diff
@@ -17,10 +17,7 @@
 package feast.serving.registry;
 
 import com.google.protobuf.Duration;
-import feast.proto.core.FeatureProto;
-import feast.proto.core.FeatureViewProto;
-import feast.proto.core.OnDemandFeatureViewProto;
-import feast.proto.core.RegistryProto;
+import feast.proto.core.*;
```
Comment (adchia) [inline java/serving/src/main/java/feast/serving/registry/RegistryRepository.java:20]: This seems like bad practice to import *?

File: java/serving/src/main/java/feast/serving/service/OnlineServingServiceV2.java [hunk 10]
```diff
@@ -182,132 +128,186 @@ public GetOnlineFeaturesResponse getOnlineFeatures(GetOnlineFeaturesRequestV2 re
...
-            this.onlineTransformationService.transformFeatures(transformFeaturesRequest);
-
-        this.onlineTransformationService.processTransformFeaturesResponse(
-            transformFeaturesResponse,
-            onDemandFeatureViewName,
-            onDemandFeatureStringReferences,
-            values,
-            statuses);
-      }
+    // populateHistogramMetrics(entityRows, featureReferences, projectName);
```
Comment (adchia) [inline java/serving/src/main/java/feast/serving/service/OnlineServingServiceV2.java:199]: Uncomment?

File: java/serving/src/main/java/feast/serving/service/OnlineServingServiceV2.java [hunk 11]
```diff
@@ -52,126 +49,77 @@
...
 
     // Get the set of request data feature names and feature inputs from the ODFV references.
-    Pair<Set<String>, List<FeatureReferenceV2>> pair =
-        this.onlineTransformationService.extractRequestDataFeatureNamesAndOnDemandFeatureInputs(
-            onDemandFeatureReferences, projectName);
-    Set<String> requestDataFeatureNames = pair.getLeft();
-    List<FeatureReferenceV2> onDemandFeatureInputs = pair.getRight();
+    List<FeatureReferenceV2> onDemandFeatureInputs =
+        this.onlineTransformationService
+            .extractRequestDataFeatureNamesAndOnDemandFeatureInputs(onDemandFeatureReferences)
```
Comment (adchia) [inline java/serving/src/main/java/feast/serving/service/OnlineServingServiceV2.java:93]: Think with your new context proto, we can clean up a bunch of onlinetransformationservice. e.g. are we still using the "getLeft()" of the extractRequestDataFeatureNamesAndOnDemandFeatureInputs?

Maybe add TODOs to clean up? I also don't really see a test we have for on demand.

Comment (adchia) [reply java/serving/src/main/java/feast/serving/service/OnlineServingServiceV2.java:93]: We can add the on demand tests after this PR merges

**Review by adchia — COMMENTED** [review]
also change the release notes to say that the java FS supports feature services now?

File: protos/feast/serving/ServingService.proto [hunk 12]
```diff
@@ -94,6 +87,11 @@ message GetOnlineFeaturesRequest {
     // A map of entity name -> list of values
     map<string, feast.types.RepeatedValue> entities = 3;
     bool full_feature_names = 4;
+
+    // Context for OnDemand Feature Transformation
```
[the remark this answers is not in the document]
Comment (pyalex) [reply protos/feast/serving/ServingService.proto:91]: To my knowledge, there was no "request_data" in protos or Python API. Not sure what you're referring to

File: java/serving/src/main/java/feast/serving/service/OnlineServingServiceV2.java (hunk 11 above)
Comment (pyalex) [reply java/serving/src/main/java/feast/serving/service/OnlineServingServiceV2.java:95]: Yeah, there were no tests. That bothers me as well, but this PR is already big enough.
ToDo added.

File: protos/feast/serving/ServingService.proto (hunk 12 above)
[the remark this answers is not in the document]
Comment (adchia) [reply protos/feast/serving/ServingService.proto:91]: This was used for the on demand feature views. You had touched some of this in the transformation logic (it used to be passed in as part of the entity row). I _think_ this is the same. 

Can you also add a new issue to migrate the python logic for this too?

**Review by adchia — APPROVED** [review]
/lgtm

**Pull request merged**